### ISSUE TITLE
User search returns invalid JSON if request caused error

### ISSUE THREAD

**Issue by 23knots** [keycloak/keycloak#20451]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

admin/api

### Describe the bug

In case of LDAP User Federation if there is an backend error while fetching users, response returned by Keycloak is not a valid JSON.  Response from the server is : 

[]{"error":"unknown_error"}

Depending on the setup array can sometimes contains user objects.

### Version

21.1.1

### Expected behavior

KC should return valid JSON for example : {"error":"unknown_error"}, without user array

### Actual behavior

KC returns array of user objects (empty or not) which is not valid JSON. Here are two responses that I am getting on two different versions and installations:

url: https://localhost:8086/auth/admin/realms/myrealm/users?briefRepresentation=true&first=0&max=20&search=*
response: []{"error":"unknown_error"}

or 

url: http://localhost:8080/admin/realms/master/ui-ext/brute-force-user?briefRepresentation=true&first=0&max=11&search=*
response: [{"id":"53111570-de45-4dbd-81f3-02923e446686","createdTimestamp":1684474467833,"username":"admin","enabled":true,"emailVerified":false,"access":{"manageGroupMembership":true,"view":true,"mapRoles":true,"impersonate":true,"manage":true},"bruteForceStatus":{"numFailures":0,"disabled":false,"lastIPFailure":"n/a","lastFailure":0}}]{"error":"unknown_error"}

### How to Reproduce?

1) Install Keycloak
2) Add LDAP/ User Federation but provide invalid connection URL
3) go to users
4) search for * 
5) check response

**Comment by jsorah** [comment]
This is likely not presenting as a major issue in the UI because although the response data is not valid JSON, the response code is still a 500 as in previous versions.

18.0.2 => `{error: "unknown_error"}`
19.0.2 => `{error: "unknown_error"}`

20.0.3, 21.1.1
```
[
    {
        "id": "6908c4e1-3bb0-486c-b0a1-958945a1f830",
        "createdTimestamp": 1684496964516,
        "username": "admin",
        "enabled": true,
        "emailVerified": false,
        "access": {
            "manageGroupMembership": true,
            "view": true,
            "mapRoles": true,
            "impersonate": true,
            "manage": true
        },
        "bruteForceStatus": {
            "numFailures": 0,
            "disabled": false,
            "lastIPFailure": "n/a",
            "lastFailure": 0
        }
    }
]{
    "error": "unknown_error"
}
```

**Comment by jsorah** [comment]
I think this warrants some investigation, adding to 22 milestone.

**Comment by patrickjennings** [comment]
This seems to be an issue with any resource method that returns a Stream. And specifically with the Quarkus distribution of Keycloak.

For these methods, there may have been evaluated and streamed content before an exception is thrown. The ExceptionMapper, KeycloakErrorHandler, will properly catch and return the correct response. But it seems with Quarkus, some of the streamed output may be flushed to the caller, prior to handling of the exception. Causing the issue we see here.

This is an interesting find. Will need to do some more investigation and see if this is something specific to Keycloak and whether this is expected with Quarkus.

**Comment by patrickjennings** [comment]
@pedroigor , if you could take a look at this issue.

I created a branch showcasing that if we evaluate the Stream to a concrete Collection in the TransactionalResponseFilter, the runtime exceptions would be handled appropriately in Quarkus.

I couldn't find the exact difference that is causing this issue in Keycloak 19+ and Quarkus. It seems like it is in how Quarkus / Vert.X is handling the Stream. Which might have changed in the upgrade of Quarkus or another related dependency. I did not see anything in our middleware code that would cause this to happen.

Let me know what you think.

**Comment by pedroigor** [comment]
@patrickjennings Thanks for the initial investigation. I'll try to spot what might be happening.

**Comment by pedroigor** [comment]
@23knots @patrickjennings After some investigation, the problem is related to how Jack/Resteasy write interceptors are handling errors when processing streams.

When processing the reference pipeline these interceptors are writing data to the response before even starting to serialize the elements so that when an exception happens, unexpected data is written to the response. For instance:

If the pipeline has a mix of users from the local database and LDAP and users from the database are the first elements, they will be serialized to the response. However, LDAP users won't as LDAP is in a failing state. In addition to that, our JAX-RS error handler will try to write to the response a proper error message. Therefore, resulting in the behavior herein described.

If a stream has only LDAP users, it will fail right at the beginning but the start/end of the JSON array is written to the response. The result is something like `[] {"error": "description"}`.

**Comment by pedroigor** [comment]
This might be a bug in Resteasy Classic and how write interceptors are handling errors when processing streams. Need to check with @geoand.

**Comment by geoand** [comment]
Hm... This is interesting. 
Just so I make sure I understand correctly, what is an example of Keycloak returning a "stream"?

**Comment by pedroigor** [comment]
An example is here https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java.

It is one of the endpoints from the issue description.

**Comment by patrickjennings** [comment]
I created an example repo of the problem and what we are doing and seeing in Keycloak:
https://github.com/patrickjennings/quarkus-response-interceptor-streaming/tree/master 

Seems that both a ExceptionMapper and a WriterInterceptor are necessary to implement as well, in order to see this behavior.

**Comment by geoand** [comment]
Ah, we are talking about actual Java Streams here. I have to admit I have no idea how these are handled in RESTEasy Classic (I am not even sure if / how we support them in RESTEasy Reactive)...

**Comment by pedroigor** [comment]
As discussed, https://github.com/quarkusio/quarkus/issues/34419.

**Comment by alechenninger** [comment]
Related to [Hynek's comment on the PR](https://github.com/keycloak/keycloak/pull/20871#discussion_r1246517752), I don't think the expected output here makes 100% sense:

> Expected behavior
> KC should return valid JSON for example : {"error":"unknown_error"}, without user array

In order to only show the exception in the response, you'd have to have buffered the response until complete without writing any to the wire, which is exactly what Streams are there to avoid. Errors during streaming can only be treated as invalid responses, or modeled as a first-class element or field in the response, because something went wrong _in the process of writing the response_. The status code is already decided by that point-it is the status of the initial cursor.

My 2c is:

1. Either accept an invalid response body is the correct semantic (though probably doesn't make sense to append the error as done now, but there's not much else you can do IMO, perhaps simply leaving the array unfinished is enough of a signal that something went wrong and the response is partial)
2. Model an error during cursor iteration as a first class citizen in the API, though it would be a breaking change. Examples could include a new top level field, written at the end (`{"data": [....], "partialResponseError": {....}}`) or an element within the stream array (`[...., {"error": {...}}]`).
3. End the array and make it appear like the response was successful, even though it's missing data

**Comment by pedroigor** [comment]
I'm not sure if I'm following your proposal because the core issue is clearly a bug in how Java Streams are handled by Quarkus. See https://github.com/quarkusio/quarkus/issues/34419.

Basically, the elements in a Java Stream should only be written to the response if all the elements are successfully serialized. Otherwise, an error message should be shown.

**Comment by alechenninger** [comment]
@pedroigor Hey thanks for the reply. What I mean it is that it is impossible to both:

- Stream responses without buffering the whole thing in memory and...
- Only return an error if an error is encountered part-way through streaming results

Using a stream but buffering the whole stream before sending (to catch any errors) would defeat the purpose of using a Stream. You might as well just use a Collection, then. I think the "expected behavior" is therefore invalid. I'll share on the Quarkus issue.

**Comment by hmlnarik** [comment]
Keycloak (or in this case perhaps Quarkus) should do the best effort to actually send a valid response. Most of the time, the error response would be sent almost instantly when there is almost no output sent across the wire.

Acceptable solution from my perspective is to buffer a part of the response (say 8 KiB). If an exception is thrown while the response is still in buffer, trash it and send an fresh error response. If the exception occurs when after this 8 KiB, then the response would become invalid. The latter scenario is unlikely though, as the exception is most likely to be thrown in the beginning of the stream processing.

One extra optimization would be to do this buffering only for `Stream`s but I don't know off-hand how straightforward or complex that would be to implement.

**Comment by pedroigor** [comment]
This issue should be closed by #10713.
